lib/tokens/TokenManager.js

```diff
@@ -0,0 +1,117 @@
+// limitations under the License.
+'use strict';
+
+const debug = require('debug')('cloudant:tokens:tokenmanager');
+const cookie = require('cookie');
+const EventEmitter = require('events');
+
+class TokenManager {
+  constructor(client, jar, sessionUrl) {
+    this._client = client;
+    this._jar = jar;
+    this._sessionUrl = sessionUrl;
+
+    this._forceRenew = true;
+    this._lastRenewResponse = null;
+    this._renewingToken = false;
+
+    this._ee = new EventEmitter().setMaxListeners(Infinity);
+  }
+
+  _autoRenew() {
+    debug('Auto renewing token now...');
+    this.forceRenew().then((response) => {
+      let maxAgeSecs = cookie.parse(response.headers['set-cookie'][0])['Max-Age'] || 3600000;
+      let delaySecs = maxAgeSecs / 2;
+      debug(`Renewing token in ${delaySecs} seconds.`);
+      setTimeout(this._autoRenew.bind(this), delaySecs * 1000).unref();
+    }).catch((error) => {
+      debug(`Failed to auto renew token - ${error}. Retrying in 60 seconds.`);
+      setTimeout(this._autoRenew.bind(this), 60000).unref();
+    });
+  }
+
+  _getToken(callback) {
+    // ** Method to be implemented by _all_ subclasses of `TokenManager` **
+    throw new Error('Not implemented.');
+  }
+
+  // Getter for `forceRenew`.
+  // - `true`  A renewal attempt will be made on the next `renew` request.
+  // - `false` No renewal attempt will be made on the next `renew`
+  //           request. Instead the last good renewal response will be returned
+  //           to the client.
+  get getforceRenew() {
+    return this._forceRenew;
+  }
+
+  // Getter for `renewingToken`.
+  // - `true`  A renewal attempt is in progress.
+  // - `false` There are no in progress renewal attempts.
+  get getRenewingToken() {
+    return this._renewingToken;
+  }
+
+  // Set `forceRenew` to `true`.
+  setForceRenew() {
+    this._forceRenew = true;
+  }
+
+  // Renew the token.
+  forceRenew() {
+    if (!this._renewingToken) {
+      this._renewingToken = true;
+      this._ee.removeAllListeners();
+      debug('Starting token renewal.');
+      this._getToken((error, response) => {
+        if (error) {
+          this._ee.emit('error', error, response);
+        } else {
+          this._ee.emit('success', response);
+          this._lastRenewResponse = response;
+          this._forceRenew = false;
+        }
+        debug('Finished token renewal.');
+        this._renewingToken = false;
+      });
+    }
+    return new Promise((resolve, reject) => {
+      this._ee.once('success', resolve);
+      this._ee.once('error', (error, response) => {
+        error.response = response;
+        reject(error);
+      });
+    });
+  }
+
+  // Renew the token if `forceRenew` is `true`. Otherwise return the last good
+  // renewal response to the client.
+  renew() {
+    if (this._forceRenew) {
+      return this.forceRenew();
+    } else {
+      return new Promise((resolve) => {
+        resolve(this._lastRenewResponse);
```

Comment:
Should we verify that `_lastRenewResponse` is not empty or null, and if it is warn or throw an error to the user?

Reply:
I initially had a test looking at this response object in the resolved promise but that's gone now so I don't think we need this. If the cookie gets put in the jar and the promise is resolved then all is good. I'll remove the `_lastRenewResponse` stuff to avoid complication.

Reply:
Oh, I think the auto renew thing uses it so perhaps it can't go entirely. I'll have a ponder.

Reply:
Ner, I was wrong. It calls the internal `_renew` so we're all good!